WalletWasabi.Fluent/Models/WebcamQrReader.cs

```diff
@@ -0,0 +1,164 @@
+using Avalonia.Media.Imaging;
+using System;
+using System.Threading.Tasks;
+using OpenCvSharp;
+using Avalonia;
+using Avalonia.Media;
+using System.Runtime.InteropServices;
+using WalletWasabi.Logging;
+using WalletWasabi.Userfacing;
+using NBitcoin;
+using Nito.AsyncEx;
+using Avalonia.Platform;
+
+namespace WalletWasabi.Fluent.Models
+{
+	public class WebcamQrReader
+	{
+		private const byte DefaultCameraId = 0;
+
+		public WebcamQrReader(Network network)
+		{
+			ScanningTaskLock = new();
+			Network = network;
+		}
+
+		public event EventHandler<WriteableBitmap>? NewImageArrived;
+
+		public event EventHandler<string>? CorrectAddressFound;
+
+		public event EventHandler<string>? InvalidAddressFound;
+
+		public event EventHandler<Exception>? ErrorOccured;
+
+		private AsyncLock ScanningTaskLock { get; set; }
+		public bool RequestEnd { get; set; }
+		public Network Network { get; }
+		public Task? ScanningTask { get; set; }
+		public bool IsRunning => ScanningTask is not null;
+
+		public async Task StartScanningAsync()
+		{
+			using (await ScanningTaskLock.LockAsync().ConfigureAwait(false))
```

Comment:
Can you explain how this is supposed to work? I think it does not work as you expect.

Reply:
The point should be that you can't call `StartScanningAsync` more than once, that could cause trouble. WIthout this, you might be able to click twice so fast it runs into an exception.
But all ideas and advices are happily taken.

Reply:
Well, but you can still call `StartScanningAsync` very fast one after another because `ScanningTask = Task.Run(() =>` just starts a task which is run basically on background. So you take a lock on L42 and get to L72 in no time really and my take is that you only protect `ScanningTask` reference change but nothing more really.

I'm not sure if this is the intended behavior.

Reply:
Added
 ```
 if (ScanningTask is { })
				{
					return;
				}
```
and moved `RequestEnd = false` out of the Task, so somehow calling Start twice, and calling Stop immediately after Start should cause no trouble.

Reply:
You set `RequestEnd` variable in multiple threads without synchronizing it. The variable should be `volatile` or under a lock.